File: src/main/java/ru/hh/typofixer/morphers/SkipChar.java
package ru.hh.typofixer.morphers;

import com.google.common.base.Function;
import com.google.common.collect.Iterables;
import ru.hh.typofixer.IntRange;
import ru.hh.typofixer.Morph;
import ru.hh.typofixer.Token;
import ru.hh.typofixer.Variant;
import ru.hh.typofixer.Word;

public class SkipChar extends Morph {
  @Override
  public Token apply(final Word input) {
    final String value = input.getValue();
    if (value.length() == 0) {
      return input;
    }
    return new Variant(Iterables.transform(IntRange.of(0, value.length() - 1),
       new Function<Integer, Word>() {
         @Override
         public Word apply(Integer offset) {
           return new Word(value.substring(0, offset).concat(value.substring(offset + 1)));
         }
       }));
  }
}

File: src/main/scala/ru/hh/typofixer/morphers/SkipChar.scala
package ru.hh.typofixer.morphers

import ru.hh.typofixer.{Variant, Word, Morph}

class SkipChar extends Morph {
  override def apply(token: Word) = {
    val value = token.string
    if (value.length == 0)
      token
    else
      new Variant(
        0 until (value.length - 1) map(idx =>
          new Word(value.substring(0, idx).concat(value.substring(idx + 1))))
      )
  }
}
